File: apps/scala/src/main/scala/machinelearning/chapter11/Metrics.scala
package machinelearning.chapter11

/** 2 値分類の混同行列。正例（見つけたいほう）を決めて、予測の当たり外れを 4 つに分けて数える。
  *
  * @param tp
  *   実際は正例で、正例と予測した件数（真陽性）
  * @param fp
  *   実際は負例で、正例と予測した件数（偽陽性）
  * @param fn
  *   実際は正例で、負例と予測した件数（偽陰性）
  * @param tn
  *   実際は負例で、負例と予測した件数（真陰性）
  */
case class ConfusionMatrix(tp: Int, fp: Int, fn: Int, tn: Int)

object ConfusionMatrix:
  /** 正解と予測を 1 件ずつ比べて数える。positive と等しいラベルを正例、それ以外を負例とする。 */
  def of[A](actual: Vector[A], predicted: Vector[A], positive: A): ConfusionMatrix =
    Metrics.requireSameSize(actual, predicted)
    actual.lazyZip(predicted).foldLeft(ConfusionMatrix(0, 0, 0, 0)) { (cm, pair) =>
      (pair._1 == positive, pair._2 == positive) match
        case (true, true)   => cm.copy(tp = cm.tp + 1)
        case (false, true)  => cm.copy(fp = cm.fp + 1)
        case (true, false)  => cm.copy(fn = cm.fn + 1)
        case (false, false) => cm.copy(tn = cm.tn + 1)
    }

/** 分類と回帰の評価指標。回帰の RMSE・MAE は第 7 章の [[machinelearning.chapter07.RegressionMetrics]] を使う。 */
object Metrics:

  /** 正解と予測の件数が同じでなければ例外を投げる。短いほうに合わせて黙って切り詰めない。 */
  private[chapter11] def requireSameSize(actual: Vector[?], predicted: Vector[?]): Unit =
    require(
      actual.size == predicted.size,
      s"正解と予測の件数が違います（正解 ${actual.size} 件、予測 ${predicted.size} 件）"
    )

  private def ratio(numerator: Double, denominator: Double): Double =
    if denominator == 0 then 0.0 else numerator / denominator

  /** 適合率。正例と予測したうち、本当に正例だった割合。 */
  def precision(cm: ConfusionMatrix): Double = ratio(cm.tp, cm.tp + cm.fp)

  /** 再現率。本当の正例のうち、正例と予測できた割合。 */
  def recall(cm: ConfusionMatrix): Double = ratio(cm.tp, cm.tp + cm.fn)

  /** F 値。適合率と再現率の調和平均。 */
  def f1Score(cm: ConfusionMatrix): Double =
    val p = precision(cm)
    val r = recall(cm)
    ratio(2 * p * r, p + r)

  /** 平均二乗誤差（MSE）。誤差の 2 乗の平均。 */
  def meanSquaredError(actual: Vector[Double], predicted: Vector[Double]): Double =
    requireSameSize(actual, predicted)
    actual.lazyZip(predicted).map((t, y) => (y - t) * (y - t)).sum / actual.size

  /** 正解率。正解と予測が一致した割合。 */
  def accuracy[A](actual: Vector[A], predicted: Vector[A]): Double =
    requireSameSize(actual, predicted)
    actual.lazyZip(predicted).count(_ == _).toDouble / actual.size

  /** 混同行列から求める指標を、正例を決めて、正解と予測から求める評価関数に変える。 */
  def classificationMetric[A](score: ConfusionMatrix => Double, positive: A): Metric[A] =
    (actual, predicted) => score(ConfusionMatrix.of(actual, predicted, positive))

File: apps/ruby/lib/getting_started_ml/chapter11/metrics.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 11 章の評価指標。
  module Chapter11
    # 2 値分類の混同行列。正例（見つけたいほう）を決めて、予測の当たり外れを 4 つに分けて数える。
    ConfusionMatrix = Data.define(:true_positive, :false_positive, :false_negative, :true_negative) do
      # 正解と予測から数える。件数が違えば ArgumentError を投げる（短いほうに合わせて黙って切り詰めない）。
      def self.of(actual, predicted, positive)
        Chapter11.require_same_size(actual, predicted)
        # 「実際が正例か」「予測が正例か」の組ごとに数える。4 通りのどれも無ければ 0 件
        counts = actual.zip(predicted).map { |truth, prediction| [truth == positive, prediction == positive] }.tally

        new(true_positive: counts.fetch([true, true], 0), false_positive: counts.fetch([false, true], 0),
            false_negative: counts.fetch([true, false], 0), true_negative: counts.fetch([false, false], 0))
      end

      # 正解率。
      def accuracy
        Chapter11.ratio(true_positive + true_negative, to_h.values.sum)
      end

      # 適合率。正例と予測したうち、本当に正例だった割合。
      def precision
        Chapter11.ratio(true_positive, true_positive + false_positive)
      end

      # 再現率。本当の正例のうち、正例と予測できた割合。
      def recall
        Chapter11.ratio(true_positive, true_positive + false_negative)
      end

      # F 値。適合率と再現率の調和平均。
      def f1_score
        Chapter11.ratio(2.0 * precision * recall, precision + recall)
      end
    end

    module_function

    # 件数が同じでなければ ArgumentError を投げる。
    def require_same_size(actual, predicted)
      raise ArgumentError, "件数が違います: #{actual.size} と #{predicted.size}" unless actual.size == predicted.size
    end

    # 分母が 0 なら 0 を返す割り算。
    def ratio(numerator, denominator)
      denominator.zero? ? 0.0 : numerator.to_f / denominator
    end

    # 混同行列から求める指標を、正例を決めて評価関数（正解と予測を受け取る lambda）に変える。
    def classification_metric(score, positive)
      ->(actual, predicted) { ConfusionMatrix.of(actual, predicted, positive).public_send(score) }
    end

    # 正解率の評価関数。第 1 章の accuracy は（予測, 正解）の順に受け取るので、順を入れ替えて包む。
    ACCURACY = ->(actual, predicted) { Chapter01.accuracy(predicted, actual) }
    # 第 7 章の RMSE と MAE は（実測, 予測）の順なので、Method オブジェクトをそのまま評価関数にできる。
    RMSE = Chapter07.method(:root_mean_squared_error)
    MAE = Chapter07.method(:mean_absolute_error)
  end
end
